dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
-            # Access rules can more easily check /proc/self.
-            normalized = os.path.join('/proc/self', os.path.relpath(file, f'/proc/{debugger.tid}'))
-        real = os.path.realpath(file)
+        # We need to project /proc/self.
+        # This might fail if the link is /proc/../proc/self or something silly like that
+        # But in such a case, the file will simply be disallowed
+        projected = self.project_proc_self(debugger, file)
+        projected_normalized = self.normalize_path(projected)
+        simplified = self.simplify_proc_pid(debugger, projected_normalized)
+        real = os.path.realpath(projected)
 
         try:
-            same = normalized == real or os.path.samefile(projected, real)
+            same = projected_normalized == real or os.path.samefile(projected, real)
         except OSError:
-            log.debug('Denying access due to inability to stat: normalizes to: %s, actually: %s', normalized, real)
-            return file, ACCESS_ENOENT
+            raise AccessFailure(
+                ACCESS_ENOENT,
+                logging.WARNING,
+                'Cannot stat: given: %s, projected to: %s, actually: %s',
+                file,
+                projected,
+                real,
+            )
         else:
             if not same:
-                log.warning(
-                    'Denying access due to suspected symlink trickery: normalizes to: %s, actually: %s',
-                    normalized,
+                raise AccessFailure(
+                    ACCESS_EACCES,
+                    logging.WARNING,
+                    'Suspected symlink trickery: given: %s, projected to: %s, actually: %s',
+                    file,
+                    projected,
                     real,
                 )
-                return file, ACCESS_EACCES
 
-        if not fs_jail.check(normalized):
-            return normalized, ACCESS_EACCES
+        if not fs_jail.check(simplified):
+            self.log_disallowed_file(syscall, file, simplified)
+            return False
+
+        if projected_normalized != real:
+            simplified_real = self.simplify_proc_pid(debugger, real)
+            if not fs_jail.check(simplified_real):
+                self.log_disallowed_file(syscall, file, simplified_real)
+                return False
 
-        if normalized != real:
-            proc_dir = f'/proc/{debugger.tid}'
-            if real.startswith(proc_dir):
-                real = os.path.join('/proc/self', os.path.relpath(real, proc_dir))
+        return True
 
-            if not fs_jail.check(real):
-                return real, ACCESS_EACCES
+    def log_disallowed_file(self, syscall: str, file: str, checked_file: str) -> None:
+        log.debug('Denied access via %s to: %s, actually checked: %s', syscall, file, checked_file)
 
-        return normalized, None
+    def get_reg_value(self, debugger: Debugger, reg: int) -> int:
+        return getattr(debugger, 'uarg%d' % reg)
```

Comment:
Please inline this, or if it really hurts your eyes create a `Debugger.uarg` that implements `[x]` as `getattr(self, 'uarg%d' % reg)`.